rfcs/0135-custom-asserts.md

```diff
@@ -0,0 +1,116 @@
+---
+feature: custom-asserts
+start-date: 2022-09-21
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Allow users to use attribute sets with a boolean attribute `success` and a string attribute `message` instead of a boolean in `assert …; …` expressions.
+
+# Motivation
+[motivation]: #motivation
+
+Since Nix is an untyped language, asserts are often needed to ensure that a function or program works correctly. However, the current assert system is unsuitable for more sophisticated error reporting that aims to inform the user as to what has happened.
+
+Consider a nixpkgs package expression that wants to validate its arguments. Currently, the best way to
+provide a custom error message is to use `assert … -> throw …; …`.
+This method has several disadvantages: Since an implication from a falsehood is always true, you are required
+to invert the condition. Additionally, since the assertion itself is not triggered by the error,
+the function of the `assert` keyword is reduced to providing an imperative shorthand for `seq`. This also means that by default,
+the error location is not printed, and there is no mention of an assert in the error message.
+Instead, expressions could use this more natural syntax:
+
+```nix
+{ foo, bar }:
+assert {
+  success = foo || bar;
+  message = "At least one of foo or bar must be set";
+};
+[ foo bar ]
+```
+
+Also consider, for instance, this simple type system:
+
+```nix
+rec {
+  assertType = type: locDescr: value:
+    if ! type.check value
+    then throw "Value at ${locDescr} was not of type ${type.name}"
+    else builtins.traceVerbose "Successfully checked type ${type.name} at ${locDescr}" value;
+  assertTypeSeq = type: locDescr: value1: value2:
+    builtins.seq (assertType type locDescr value1) value2;
+  intType = {
+    check = builtins.isInt;
+    name = "integer";
+  };
+  attrsOfType = subType: {
+    check = value:
+      builtins.isAttrs value
+      && builtins.all subType.check (builtins.attrValues value);
+    name = "attribute set of ${subType.name}";
+  };
+}
+```
+
+Users of this system would be forced to forgo the convenience of imperative-style `assert …; …` in favor of `seq`-like syntax in order to benefit from improved type errors. With this change, no longer! A variant `isType` function could be declared:
```

Comment:
You can kind of have your cake and eat it, too, with a custom system. Such one has actually been added to nixpkgs by @roberth:

```
lib.throwIfNot cond1 "msg1"
lib.throwIfNot cond2 "msg2"
stdenv.mkDerivation { /* … */ }
```

This relies on a neat trick that you can return the identity function from a function, causing anything further in the source file to act as if that `lib.throwIfNot` and its first two arguments hadn't been there.

Reply:
oh that’s neat! I still think `assert …;` is more idiomatic

Reply:
See this prior discussion https://github.com/NixOS/nixpkgs/issues/154292

Nitpick: Idioms evolve from a language and its use, not the other way around. We're designing a language feature, so existing idioms are relevant but of secondary importance.

Reply:
that’s a nice way to view it! I rephrase: I think my suggestion would make it more natural, easier to understand and more distinctive. It uses the syntax specifically for asserts and only for that, instead of co-opting returning identity sometimes.

Reply:
In my opinion, these other idioms do not demonstrate that this change is not needed, but demonstrate that there is a lack of unified assertion functionality, which has left the conciseness of the dedicated `assert` syntax behind.

Reply:
If we had `$` like Haskell, we wouldn't miss the `assert ..;` syntax so much because we wouldn't need more parentheses either way.

Reply:
`$` keeps getting brought up, *but* it is an even more invasive change and would introduce more unfamiliar syntax to a language non FP programmers already struggle with conceptually…

Reply:
Yeah, IMO `assert` is a more user-friendly idiom. And it already exists in the language!

Reply:
That's little a bit unfair: for anybody with no lazy FP experience, syntax will be the least of their problems.  No debugger, no `printf()`, and brain-meltingly weird stack traces.

But yeah, language stability is precious.